alternator/executor.cc

```diff
@@ -507,17 +504,21 @@ static rjson::value fill_table_description(schema_ptr schema, table_status tbl_s
             rjson::add(table_description, "GlobalSecondaryIndexes", std::move(gsi_array));
         }
     }
-    // Use map built by describe_key_schema() for base and indexes to produce
-    // AttributeDefinitions for all key columns:
-    rjson::value attribute_definitions = rjson::empty_array();
-    for (auto& type : key_attribute_types) {
-        rjson::value key = rjson::empty_object();
-        rjson::add(key, "AttributeName", rjson::from_string(type.first));
-        rjson::add(key, "AttributeType", rjson::from_string(type.second));
-        rjson::push_back(attribute_definitions, std::move(key));
+    if (tbl_status != table_status::deleting) {
+        // Add base table's KeySchema and collect types for AttributeDefinitions:
+        executor::describe_key_schema(table_description, *schema, key_attribute_types);
```

Comment:
Please move the declaration
```C++
std::unordered_map<std::string,std::string> key_attribute_types;
```
which you left in its original place to the new location here. Unless I'm missing something, there is no reason why the definition needs to stay many lines up

Reply:
This declaration is being used earlier in the code on string 488 (`executor::describe_key_schema(view_entry, *vptr, key_attribute_types);` ), so i can't move it lower.

Reply:
Oh, I missed that...
But this brings another question: It's strange that the code above unconditionally (also in delete) adds the views' information in key_attribute_types, which won't be used, but only for the base table we don't. The loop above doesn't have the new `tbl_status != table_status::deleting` check you added here. I'm worried that if we DeleteTable a table which has a GSI or LSI, maybe we'll unnecessarily add its schema to the result.

I think we're missing a test for DeleteTable that would check the result of a DeleteTable of a table with GSI - I think the DeleteTable-result test we have in test_table.py doesn't have a GSI in its table, and we should probably add one?

Reply:
i'll add the != deleting condition to the loop, and then try to create a DeleteTable test for GSI/LSI case

Reply:
I just want to clarify, where exactly to put this condition?

Reply:
There's a `if (!t.views().empty()) {` containing a loop of all the views. Do we need this for DeleteTable? Does DeleteTable return Local/GlobalSecondaryIndexes or doesn't it? If it doesn't, this whole if can be skipped. If it does, perhaps just something inside it which prints the schemas can be skipped.

This is why I always try to write the test before the development - to see what I *really* need to code. I understand you're having trouble with AWS. If you want, I can write a test for this and you'll just need to make the test pass with this code. Shall I write this test?

Reply:
I'm writing this test, should be ready within an hour :-)

Reply:
> I'm writing this test, should be ready within an hour :-)

OK :slightly_smiling_face:

Reply:
My test (I'll include it below) confirms that the GlobalSecondaryIndexes and LocalSecondaryIndexes should also be missing for DeleteTable, so perhaps the entire if() above should just be conditional on the state not being "deleting". I'll post the test below in a minute.